lib/Transforms/LLZKPolyLoweringPass.cpp

```diff
@@ -0,0 +1,492 @@
+      auto newConst = builder.create<FeltConstantOp>(constOp.getLoc(), constOp.getValue());
+      rebuildMemo[val] = newConst.getResult();
+      return newConst.getResult();
+    }
+
+    llvm::errs() << "Unhandled expression kind in rebuildExprInCompute: " << val << "\n";
+    llvm_unreachable("Unsupported op in rebuildExprInCompute");
+  }
+
+  void runOnOperation() override {
+    ModuleOp moduleOp = getOperation();
+
+    // Validate degree parameter
+    if (maxDegree < 2) {
+      moduleOp.emitError() << "Invalid max degree: " << maxDegree << ". Must be >= 2.";
+      signalPassFailure();
+      return;
+    }
+
+    moduleOp.walk([&](StructDefOp structDef) {
+      FuncDefOp constrainFunc = structDef.getConstrainFuncOp();
+      FuncDefOp computeFunc = structDef.getComputeFuncOp();
+      if (!constrainFunc) {
+        moduleOp.emitError() << "Struct: " << structDef.getName()
+                             << " doesn't have a constrain func";
+        signalPassFailure();
+        return;
+      }
+
+      if (!computeFunc) {
+        moduleOp.emitError() << "Struct: " << structDef.getName() << " doesn't have a compute func";
+        signalPassFailure();
+        return;
+      }
+
+      DenseMap<Value, unsigned> degreeMemo;
+      DenseMap<Value, Value> rewrites;
+      SmallVector<AuxAssignment> auxAssignments;
+
+      // Lower equality constraints
+      constrainFunc.walk([&](EmitEqualityOp constraintOp) {
+        Value lhsExpr = constraintOp.getOperand(0);
+        Value rhsExpr = constraintOp.getOperand(1);
+        unsigned degreeLhs = getDegree(lhsExpr, degreeMemo);
+        unsigned degreeRhs = getDegree(rhsExpr, degreeMemo);
+
+        if (degreeLhs > maxDegree) {
+          Value loweredExpr = lowerExpression(
+              lhsExpr, maxDegree, structDef, constrainFunc, degreeMemo, rewrites, auxAssignments
+          );
+          constraintOp.setOperand(0, loweredExpr);
+        }
+        if (degreeRhs > maxDegree) {
+          Value loweredExpr = lowerExpression(
+              rhsExpr, maxDegree, structDef, constrainFunc, degreeMemo, rewrites, auxAssignments
+          );
+          constraintOp.setOperand(1, loweredExpr);
+        }
+      });
+
+      // The pass doesn't currently support EmitContainmentOp as it depends on
+      // https://veridise.atlassian.net/browse/LLZK-245 being fixed Once this is fixed, the op
+      // should lower all the elements in the row being looked up
+      constrainFunc.walk([&](EmitContainmentOp containOp) {
+        moduleOp.emitError() << "EmitContainmentOp is unsupported for now in the lowering pass";
+        signalPassFailure();
+        return;
+      });
+
+      // Lower function call arguments
+      constrainFunc.walk([&](CallOp callOp) {
+        if (callOp.calleeIsStructConstrain()) {
+          SmallVector<Value> newOperands = llvm::to_vector(callOp.getArgOperands());
+          bool modified = false;
+
+          for (auto [idx, arg] : llvm::enumerate(newOperands)) {
```

Comment:
Could you just use `for (Value &arg : newOperands)...` and then `arg = ...` in the loop?